### ISSUE TITLE
stack-exec -- pacman and environments on windows.

### ISSUE THREAD

**Issue by cmoore** [commercialhaskell/stack#1182]
* Recent (last 24 hours) stack built with `stack upgrade --git` in path.
* `STACK_ROOT` set to `c:\stack`
* Fresh-ish install - `AppData\Local\Programs\stack` removed and STACK_ROOT cleared out.

> `stack setup` ;; Does what you'd expect.
> `stack exec -- pacman -Sy` ;; Otherwise, some of the packages later won't install.
> `stack exec -- pacman -Sy mingw-w64-x86_64-webkitgtk3 mingw-w64-x86_64-pkg-config` ;; lots of dependencies.
> `stack exec -- pkg-config --list-all`
> Executable named pkg-config not found on path: [big honkin' huge path]

If you go into the `msys2-DATE` folder and run `mingw64_shell.bat`, and then run `pkg-config --list-all`, all of the packages show up and taunt you mercilessly.  If instead, you run `msys2_shell` you get the command not found error above.

For the sake of completeness, I installed 32bit versions with
> `stack exec -- pacman -Sy mingw-w64-x86_64-webkitgtk3 mingw-w64-x86_64-pkg-config`

just to make sure stack wasn't defaulting to 32bit for some reason, but I got the same result.

I even tried downloading a couple of the dependencies, gio for instance, and removing the `pkg-config` sections and trying a `stack init` and then `stack build`, but of course it couldn't find the headers it was looking for.

It would seem from reading [this](http://sourceforge.net/p/msys2/wiki/MSYS2%20introduction/), that mixing environments is a pretty bad idea and my few attempts at copying things around to just make it work seemed to confirm that.

Is there a way to choose which environment (msys2, mingw64, and mingw32) `stack exec --` uses, or am I simply not doing it right?

**Comment by cmoore** [comment]
Most of the above was predicated on the idea that to interact with mingw64, I had to go through `stack exec --` which isn't necessarily the case.

If you add the relevant environment variables, in this specific case `PKG_CONFIG_PATH`, and add `mingw64/bin` to your path, things get closer to building.  I'm stuck for the moment because something is inserting mingw64 relevant paths into cairo's `'include-dirs`, which makes `ghc-pkg` choke, but I can't say for sure if it's stack doing this, so I'm going to close this issue at least for now.

**Comment by snoyberg** [comment]
Actually, Stack currently _does_ modify the include-dirs automatically, and
perhaps this is in error. You can see the code at:

https://github.com/commercialhaskell/stack/blob/master/src/Stack/Setup/Installed.hs#L119

Does playing around with that help your situation?

**Comment by cmoore** [comment]
I'm looking through the code for something that affects the haddock docs exclusively.  The best I was able to do messing with [line 124](https://github.com/commercialhaskell/stack/blob/master/src/Stack/Setup/Installed.hs#L124) was making it choke on the includes, which I guess isn't surprising.

One thing I found interesting is if you change the build flags, in my case by building with a version of stack that had lines 123 to 126 commented out, when you go to build cairo, it reports which flags were changed as always, but the paths it shows are the full C:\ path, and not the mingw64 relative path that is shown in the ghc-pkg error.

I'll reproduce these errors and link a gist of them in case there's something in them I'm missing.

**Comment by cmoore** [comment]
[The gist of it](https://gist.github.com/cmoore/822d760c350f78efcc8a)

**Comment by cmoore** [comment]
Oh, one thing I forgot to add is that `pkg-config` doesn't report paths relative to the msys2 install, which I was somewhat surprised by.

```
C:\projects\stack>stack exec -- pkg-config --cflags cairo
-mms-bitfields -IC:/Users/clint/AppData/Local/Programs/stack/x86_64-windows/msys2-20150512/mingw64/include/cairo  ...
```

**Comment by cmoore** [comment]
For those trying to shave this (massive) yak, who might find this page in their search....

[https://github.com/leksah/leksah/blob/master/Dockerfile#L52](https://github.com/leksah/leksah/blob/master/Dockerfile#L52)

The above link changes the include cflags in the `fontconfig.pc` file to a dos-relative pathname.  I had a local install of msys2 so I had to change them to:

`-I\\msys64\\mingw64\\include\\freetype2`

(I swear I tried this)

You'll also want to build with the head of the gtk2hs repository because [there's a bug in the CFLAGS and C99 print format](https://github.com/gtk2hs/gtk2hs/issues/110).  If a new gtk2hs-buildtools shows up on hackage newer than today (Oct 20, '15), it'll have the fix in it.

Oh, and you'll need to change all occurrences of `-pthread` to `-lpthread` in the pkgconfig files in msys2.  The version of GCC that ships with ghc 7.10.2 doesn't recognize it.
